### ISSUE TITLE
What is the proper way to use entity initializer ?

### ISSUE THREAD

**Issue by graphicsxp** [Breeze/breeze.bridge2.angular#2]
I'm migrating a JS project to TypeScript and I'm having a hard time figuring out how to use the entity initializer. This is what I have so far (only the relevant code) : 

in registration-helper.ts  :

` metadataStore.registerEntityTypeCtor('PhysicalFile', PhysicalFile, PhysicalFile.physicalFilePostInitializer);`

in physical-file.ts (my model):

```

export class PhysicalFile {

    constructor(public entityManagerService: EntityManagerService) {    }

    static physicalFilePostInitializer(physicalFile: PhysicalFile) {
        if (physicalFile.id === DataType.Guid.defaultValue) {
            physicalFile.documentFormat = physicalFile.entityManagerService.callAMethod( );
        }
    }
}
```

This does not work because the entityManagerService is never injected by breeze as it creates entities like this :

` var instance = new aCtor();`

What would be the correct approach to solve this scenario ?  One way would be to have the entityManagerService as static but it also is injected its own dependencies so they would have to be static too, which is not what I want.

**Comment by marcelgood** [comment]
Entities shouldn't really have dependencies. They are just data objects. Logic should generally be moved to a service that knows how to act on the data. To create entities and initialize them, we tend to use factories. Here's an example. In this case the factory is even asynchronous, so we can query for additional data that we need to initialize the entity. That's not really what I would recommend anymore at this point. We tend to create synchronous factories these days, so the retrieval of data that is required should occur outside of the factory. 

https://github.com/Breeze/temphire.angular/blob/master/TempHire/app/resourcemgt/resource-mgt-unit-of-work.ts#L24

**Comment by graphicsxp** [comment]
Thank you for your reply.

I've checked the sample. The fact that all entities initializers are empty makes me wonder what is its purpose ?
In my JavaScript app, I used the initializers to set default values on my entities. This sometimes implied to fetch data from the cache (the required data were loaded at startup). I just had to inject the entityManager factory in my entity. 
I can no longer do that with TypeScript due to the static nature of the initializer. 

So I guess I should move this code to the repository factory where I create and retrieve entities.

**Comment by marcelgood** [comment]
The initalizer function is intended to further initialize the entity after construction. The main difference between the constructor and the initializer is that if you add properties in the ctor, they are seen and managed by Breeze, whereas properties added in the intitializer will not be touched by Breeze at all. 

What I don't quite understand is why these properties you are setting are not persisted. It's a bit dangerous to set properties when querying an entity as that could leave the entity already in modified state. That usually doesn't make sense, since a freshly queried entity is normally unchanged since you just read it from its persistent store. 

The initializer function doesn't have to be a static on the entity. That's just what we chose in our TS entity generator. You can modify the generator and generate the initializer in a different way. 

You can certainly initialize properties with values from the cache in the initializer. You can access the `EntityAspect` from the entity, which also contains the `EntityManager`, so you can get at the cache from inside the entity.

**Comment by graphicsxp** [comment]
Ok after trying a few things I ended up with the conclusion that I must initialize some-client side properties in the initializer and they require dependencies to be injected in my entities (not many but a few). 

I'm stuck because I don't see how to get around the static registrationHelper. You said it is possible to modify the generator, but even without going that far, could you show me a simple use-case with a non static registration helper ? 

Here's my entity :  

```
    export class PhysicalFile {

    constructor(public entityManagerService: EntityManagerService) {    }

  configureMetadataStore(metadataStore: MetadataStore) {
        metadataStore.registerEntityTypeCtor('PhysicalFile', PhysicalFile, this.physicalFilePostInitializer);
    }
    physicalFilePostInitializer(physicalFile: PhysicalFile) {
        if (physicalFile.id === DataType.Guid.defaultValue) {
            physicalFile.documentFormat = physicalFile.entityManagerService.callAMethod( );
        }
    }
}

```
So where and when should I call the `configureMetadataStore`  method  ?

**Comment by marcelgood** [comment]
Well, you wouldn't make the initializer non-static, but you would change how you generate the RegistrationHelper, so you can inject your service into the RegistrationHelper and have it pass it to the initilizer. So for example, you would change the signature of the initializer first.

```
    /// [Initializer]
    static initializer(entity: StaffingResource, entityManagerService: EntityManagerService) { }
```

Then, depending on the framework you are using you can generate the RegistrationHelper like this. In this example I'm assuming you are using Angular, so I would generate the RegistrationHelper as an injectable Angular service and inject my dependency in the constructor and then make the register function non-static and change how the initializer is being called.

```
@Injectable()
export class RegistrationHelper {

    constructor(public entityManagerService: EntityManagerService) {}

    register(metadataStore: MetadataStore) {
        metadataStore.registerEntityTypeCtor('StaffingResource', (entity: StaffingResource) => StaffingResource.initializer(entity, this.entityManagerService));
    }
}
```

Then I can inject the RegistrationHelper in my app and call the register method.

**Comment by marcelgood** [comment]
Here are the corresponding changes to the generator. You only have to change the handlebar templates.  You'll have to put the actual path to your EntityManagerService module.

register.template.txt:
```
import { Injectable } from '@angular/core';
import { MetadataStore } from 'breeze-client';
import { EntityManagerService } from 'path-to-entitymanagerservice';

{{#modules}}import { {{path}} } from './{{moduleName}}';
{{/modules}}

@Injectable()
export class RegistrationHelper {

    constructor(public entityManagerService: EntityManagerService) {}

    register(metadataStore: MetadataStore) {
        {{#modules}}
        metadataStore.registerEntityTypeCtor('{{entityType.shortName}}', {{entityType.shortName}}{{#entityType.initializerFn}}, (entity: {{../entityType.shortName}}) => {{../entityType.shortName}}.{{.}}{{/entityType.initializerFn}}(entity, this.entityManagerService));
        {{/modules}}
    }
}

```

entity.template.txt
```
import { EntityManagerService } from 'path-to-entitymanagerservice';
{{#if baseClass}}
import { {{baseClass}} } from './{{baseClassModuleName}}';
{{/if}}
{{#imports}}import { {{path}} } from './{{moduleName}}';
{{/imports}}

/// <code-import> Place custom imports between <code-import> tags
{{{codeimport}}}
/// </code-import>

export class {{shortName}} {{#baseClass}}extends {{.}} {{/baseClass}}{

    /// <code> Place custom code between <code> tags
    {{{code}}}
    {{#generateInitializer}}
    /// [Initializer]
    static {{initializerFn}}(entity: {{shortName}}, entityManagerService: EntityManagerService) { }
    {{/generateInitializer}}
    /// </code>

    // Generated code. Do not place code below this line.
    {{#properties}}
    {{name}}: {{dataType}};
    {{/properties}}
}
```

**Comment by graphicsxp** [comment]
Thanks a lot for this. 

However I've realised that injecting entity dependencies in the registrationHelper would soon become unmanageable and not so clean. I have a lot of entities and sometimes the need a dependency, sometimes not. 

So instead, I've kept the registration helper as generated, and I use the Angular Injector  provider to resolve the dependencies in the entity (when required). It works fine and at least I keep the registration untouched. 

I have another issue with TypeScript related to Validation, but I'll open another issue. Thanks :)
